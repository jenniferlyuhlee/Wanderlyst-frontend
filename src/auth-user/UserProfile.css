.Profile-title{
    text-align: center;
}
.Profile-header{
    padding: 0 3rem;
    display: flex;
    align-items: center;
    justify-content: space-between;
}
.Profile-header-group{
    display: flex;
    align-items: center;
}
.Profile-img{
    height: 18vw;
    width: 18vw;
    border-radius: 50%;
}
.Profile-info{
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 2rem 3rem;
}

@media (max-width: 800px){
    .Profile-header{
        flex-direction: column;
        padding: 0;
    }
    .Profile-header-group{
        flex-direction: column;
    }
    .Profile-img{
        height: 150px;
        width: 150px;
    }
    .Profile-info{
        margin-top: 1rem;
        padding: 0;
    }
}

.Profile-separator{
    margin: 2rem 0 0 0;
}

@media (max-width: 1200px){
    .Profile-separator{
        margin: 0;
    }
}

.Profile-add-itin{
    color: rgba(102, 209, 149);
}

.Profile-itins-menu{
    display: flex;
    justify-content: space-around;
}

.itin-menu{
    border-radius: 0;
    padding: 0.5rem 0;
    width: 100%;
    background-color: white;
    border: 0;
    border-bottom: solid white 3px;
    color: #888;
    transition: color, 0.2s ease-out;
}
.itin-menu:hover{
    color: #555555;
}

.active-tab{
    pointer-events: none;
    cursor: not-allowed;
    color: rgba(102, 209, 149);
    border-bottom: 3px solid;
}

.bi{
    font-size: 3rem;
}

.bi-geo-alt-fill{
    font-size: 1rem;
}