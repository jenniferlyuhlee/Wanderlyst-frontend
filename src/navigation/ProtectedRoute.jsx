import React, { useContext } from "react";
import { Navigate } from "react-router-dom";
import UserContext from "../auth-user/UserContext";

/** ProtectedRoute element
 * Redirects unauthorized(anon) users to login page.
 */

function ProtectedRoute({element}){
    const {currUser} = useContext(UserContext);
    if(!currUser){
        return <Navigate to = "/login" />
    }
    return element;
}

export default ProtectedRoute;