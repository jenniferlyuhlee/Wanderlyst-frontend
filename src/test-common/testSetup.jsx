import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };