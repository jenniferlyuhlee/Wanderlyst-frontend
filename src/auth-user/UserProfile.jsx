import React, { useState, useEffect, useContext} from "react";
import { NavLink, useParams } from "react-router-dom";
import WanderlystApi from "../utils/api";
import UserContext from "./UserContext";
import dateConvert from "../helpers/dateConvert";
import ItinCard from "../itineraries/ItinCard";
import Loading from "../shared/Loading";
import "./UserProfile.css"

/** User Profile Page Component
 * Displays user details and associated itineraries (created/likes)
 */

function UserProfile(){
    const { currUser, likes} = useContext(UserContext);
    const { username } = useParams();

    // state set to null to use loading spinner
    const [user, setUser] = useState(null);
    // error state for invalid param user username
    const [error, setError] = useState(false);
    // toggle display of itineraries/likes
    const [showItins, setShowItins] = useState(true);

    useEffect(function getUserDetails() {
        async function getUser(){
            try{
                setUser(await WanderlystApi.getUser(username));
            }
            catch(err){
                console.error("User not found");
                setError(true);
            }
        }
        getUser();
    }, [username, likes]);

    // toggle itineraries / likes
    function toggleDisplay(){
        setShowItins(state => !state)
    }

    if(error) return <p className="NotFound"><i>Sorry, this user could not be found.</i></p>

    if(!user) return <Loading />

    return(
        <div className = "container pt-2 mb-5 body-cont">
            <h1 className = "Profile-title">@{user.username}</h1>
            <div className = "Profile-header">
                <div className = "Profile-header-group">
                    <img className = "Profile-img" src = 
                    {user.profilePic ? `${user.profilePic}`
                        : "/src/assets/default_profile.jpg"
                    }
                    />
                    <div className = "Profile-info">
                        <h4 className = "Profile-name">{user.firstName} {user.lastName}</h4>
                        {user.location ? 
                            <p>
                                <i className="bi bi-geo-alt-fill"></i> {user.location}
                            </p> 
                        : null}
                        {user.bio ? <p>{user.bio}</p> : null}
                        <small className = "mt-2"><i>Joined {dateConvert(user.createdAt)}</i></small>
                        {currUser.username === user.username ? 
                        <NavLink to={`/users/profile/edit`}
                            className = "mt-3 btn btn-sm rounded-pill btn-secondary">
                            Edit Profile
                        </NavLink>
                        : null}
                    </div>
                </div>
                {currUser.username === user.username ? 
                    <NavLink to="/itineraries/new" className="Profile-add-itin">
                        <i className="bi bi-plus-circle-fill"></i>
                    </NavLink>
                : null}
            </div>
            <hr className = "Profile-separator"/>
            <div>
                <div className = "Profile-itins-menu">
                    <button onClick={toggleDisplay}
                    className = {`itin-menu ${showItins? "active-tab" : ""}`}>
                        <i className = {`bi bi-map${showItins? "-fill" : ""}`}></i>
                    </button>
                    <button onClick={toggleDisplay}
                    className = {`itin-menu ${showItins? "" : "active-tab"}`}>
                        <i className = {`bi bi-bookmark-heart${showItins? "" : "-fill"}`}></i>
                    </button>
                </div>
                {showItins ?
                <div className = "User-itins m-4">
                    <div className = "mt-2">
                    {user.itineraries.length > 0 ?
                         user.itineraries.map(i => (
                            <ItinCard key = {i.id} itinerary={i}/>
                        ))
                    :
                        <div className = "text-center">
                            <i>
                                {currUser.username === user.username? "Start sharing itineraries!" : `${user.username} has no itineraries yet.`}
                            </i>
                        </div>
                    }
                    </div>
                </div>
                :
                <div className = "userLikes m-4">                
                    <div className = "mt-2">
                    {user.likes.length > 0 ?
                        user.likes.map(i => (
                            <ItinCard key = {i.id} itinerary={i}/>
                        ))
                    :
                    <div className = "text-center">
                        <i>
                            {currUser.username === user.username? "Start exploring and liking itineraries!" : `${user.username} has no likes yet.`}
                        </i>
                    </div>
                    }
                    </div> 
                </div>
                }
            </div>
        </div>
    )
}

export default UserProfile;