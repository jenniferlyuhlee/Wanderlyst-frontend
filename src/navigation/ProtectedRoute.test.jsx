import React from "react";
import { render } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter} from "react-router-dom";
import ProtectedRoute from "./ProtectedRoute";
import UserProfile from "../auth-user/UserProfile";
import { expect, describe, test, vi} from "vitest";

describe("ProtectedRoute component", () => {
    test("renders without crashing", () => {
        render(
            <MemoryRouter>
                <UserProvider>
                    <ProtectedRoute />
                </UserProvider>
            </MemoryRouter>)
    });
    test("matches snapshot", () => {
        const {asFragment} = render(
            <MemoryRouter>
                <UserProvider>
                    <ProtectedRoute/>
                </UserProvider>
            </MemoryRouter>)
        expect(asFragment()).toMatchSnapshot();
    });
});