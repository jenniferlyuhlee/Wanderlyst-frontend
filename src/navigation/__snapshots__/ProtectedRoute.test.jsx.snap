// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProtectedRoute component > matches snapshot 1`] = `<DocumentFragment />`;
